File: odroid_test/apple.cpp
#include <opencv2/core.hpp>
#include <opencv2/videoio.hpp>
#include <opencv2/highgui.hpp>
#include "opencv2/imgproc.hpp"
#include <iostream>
#include <stdio.h>

using namespace cv;
using namespace std;

// the lower and upper boundaries of the HSV color space
const int hsv_lower[3] = {5, 92, 0};
const int hsv_upper[3] = {180, 255, 150};

void getImageMask(const cv::Mat& hsv, cv::Mat& mask) {
	// smooths the image
	cv::Mat blur;
	cv::GaussianBlur(hsv, mask, Size(5, 5), 0, 0 );
	cv::inRange(mask, Scalar(hsv_lower[0], hsv_lower[1], hsv_lower[2]),
				Scalar(hsv_upper[0], hsv_upper[1], hsv_upper[2]), mask);
	// Element = 0: Rect - 1: Cross - 2: Ellipse
	cv::Mat opening_kernel = getStructuringElement(/*Element*/2, Size(4,4));
	cv::morphologyEx(mask, mask, cv::MORPH_OPEN, opening_kernel);
	cv::Mat closing_kernel = getStructuringElement(/*Element*/2, Size(10,10));
	cv::morphologyEx(mask, mask, cv::MORPH_CLOSE, closing_kernel);
}

int main(int, char**)
{
    cv::Mat frame;
    cv::Mat hsv, mask;
    std::vector<Vec4i> hierarchy;
	std::vector<std::vector<cv::Point> > contours;

    frame = cv::imread("DSCN1771.JPG");

    // convert frame to the HSV color space
    cv::cvtColor(frame, hsv, CV_BGR2HSV);

    // get mask
    getImageMask(hsv, mask);

    // find contours in the mask 
    cv::findContours(mask, contours, hierarchy,
                RETR_EXTERNAL, CHAIN_APPROX_SIMPLE, Point(0, 0) );

    for(int i = 0; i< contours.size(); i++) {
        Point2f center;
        float radius;
        cv::minEnclosingCircle(contours[i], center, radius);
        // calculate the circularity of the contour
        double area = cv::contourArea(contours[i]);
        double perimeter = cv::arcLength(contours[i], true);
        double circularity = 4*/*PI*/3.14*area/perimeter/perimeter;
        if (radius > 30 && circularity > 0.)
            cv::circle(frame, center, radius, Scalar(0, 255, 255), 2);
    }
    // the camera will be deinitialized automatically in VideoCapture destructor
    return 0;
}

File: odroid_test/apple.py
import numpy as np
import argparse
import cv2
import math

def get_image_mask(hsv_img, lower_thres, upper_thres):
  # blurring smooths the image and allows for slightly more accurate circle detection, as part of the next step includes edge detection
  mask = cv2.GaussianBlur(hsv_img, (5, 5), 0)
  # construct a mask for the hsv image, then perform closing to remove 'salt and pepper'
  mask = cv2.inRange(mask, lower_thres, upper_thres)
  # mask = cv2.dilate(mask, None, iterations=2)
  opening_kernel = cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (4,4))
  mask = cv2.morphologyEx(mask, cv2.MORPH_OPEN, opening_kernel)
  closing_kernel = cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (10,10))
  mask = cv2.morphologyEx(mask, cv2.MORPH_CLOSE, closing_kernel)

  # mask = cv2.erode(mask, None, iterations=2)
  
  return mask

def main():
  # define the lower and upper boundaries of the HSV color space
  hsv_lower = (5, 92, 0)
  hsv_upper = (180, 255, 150)

  # Load an image in RGB 
  frame = cv2.imread('DSCN1771.JPG',1)
  # Rescale image to falf x and y
  # frame = cv2.resize(frame, (0,0), fx=0.5, fy=0.5) 

  # convert frame to the HSV color space
  hsv = cv2.cvtColor(frame, cv2.COLOR_BGR2HSV)
 
  # get mask
  mask = get_image_mask(hsv, hsv_lower, hsv_upper)

  # find contours in the mask 
  contours = cv2.findContours(mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)[-2]
 
  # only proceed if contour was found
  for contour in contours:
    ((x, y), radius) = cv2.minEnclosingCircle(contour)
    #calculate the circularity of the contour
    area = cv2.contourArea(contour)
    perimeter = cv2.arcLength(contour,True)
    if perimeter>0:
      circularity = 4*math.pi*area/perimeter/perimeter
  

if __name__ == "__main__":
    main()